sbi/simulators/simutils.py

```diff
@@ -2,82 +2,480 @@
+            (i.e. a batch), not a prior on a multi-dimensional parameter.
+
+            Please use utils.torchutils.BoxUniform if you'd rather put a prior on a 
+            multi-dimensional parameter.
+            """
+        )
 
-def get_simulator_name(simulator_fun, name=None) -> str:
-    """Get or set name of the simulator. 
+
+def check_for_possibly_batched_observations(true_observation: Tensor):
+    """Raise ValueError if dimensionality of data doesn't match requirements.
     
-    Returns function name if no name is given. 
+    The current requirements are that is 2D with a leading batch dimension of size 1.
+    That is, neither multiple, batched observations are supported, 
+    nor a multidimensional observation e.g., an image.
+    """
+
+    if true_observation.squeeze().ndim > 1:
+        raise ValueError(
+            """`true_observation` has D>1 dimensions. SBI interprets the leading 
+                dimension as a batch dimension, but it *currently* only processes 
+                a single observation, i.e. the first element of the batch.
+                
+                For example:
+                
+                > true_observation = [ [1,2,3], [4,5,6] ] 
+                
+                is interpreted as two vector observations, only the first of which 
+                is currently used to condition inference.
+                
+                Use rather:
+                
+                > true_observation = [ [[1,2,3], [4,5,6]] ]
+                > true_observation = [ [1], [2], [3]]
+                
+                if your single observation is matrix-shaped or scalar-shaped.
+                
+                Finally:
+                
+                > true_observation = [1]
+                > true_observation = [1, 2, 3]
+                
+                will be interpreted as one scalar observation or one vector observation
+                and don't require wrapping (unsqueezing).
+                """
+        )
+
+
+def check_prior_methods(prior):
+    """Check whether the prior has methods .sample and .log_prob. 
     
-    Arguments:
-        simulator_fun {function} -- simulator function taking parameter batch as only argument, return data. 
+    Raises: 
+        AttributionError: if either of the two methods doesnt works as expected. 
+    """
+
+    # Sample a batch of two parameters to check batch behaviour > 1.
+    num_samples = 2
+    # Check .sample and log_prob, keep for later use.
+    try:
+        theta = prior.sample((num_samples,))
+    except AttributeError:
+        raise AttributeError(
+            "Prior needs method .sample((num_samples, )). Consider using a "
+            "pytorch Distribution"
+        )
+    try:
+        log_probs = prior.log_prob(theta)
+    except AttributeError:
+        raise AttributeError(
+            "Prior needs method .log_prob(values). Consider using a pytorch Distribution"
```

Comment:
I see what you mean, but I think I need the explicit test to make sure the two methods work like `Distribution` methods, i.e., have the right signature, accept `Tuple` as argument etc. Do you agree?

Reply:
No, I would agree if you would catch the errors resulting from e.g. shape or type mismatch. As it is you only catch `AttributeError`.

Reply:
yes, you are right. This needs more error handling. But using `hasattr` would lead to more assertions because it evaluates to `bool`. I think it is more readible if I just catch different exceptions and give the corresponding error message.

Reply:
Yes, you are right.